### PR TITLE
SNOW-1637101: Support binary operations between two timedeltas.

### PR DESCRIPTION
**Pull request by sfc-gh-mvashishtha**
SNOW-1637101

### CODE REVIEW

**Review by sfc-gh-nkrishna — APPROVED** [review]
LGTM! Nice work

File: tests/notebooks/modin/MIMICHealthcareDemo.ipynb
Comment (sfc-gh-azhan) [inline tests/notebooks/modin/MIMICHealthcareDemo.ipynb:395]: Can you double check `TimeSeriesTesting.ipynb`? We may have some improvements there too.

File: tests/integ/modin/test_timedelta_ops.py
Comment (sfc-gh-azhan) [inline tests/integ/modin/test_timedelta_ops.py:34]: nit: use a better name

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:17375]: `equal_null` is used 20 times under `src/snowflake/snowpark/modin/`. I believe we need to update all of them. Am I right?

File: tests/notebooks/modin/MIMICHealthcareDemo.ipynb
Comment (sfc-gh-mvashishtha) [reply tests/notebooks/modin/MIMICHealthcareDemo.ipynb:395]: That notebook is blocked by a features that are not related to timedelta:

- [DatetimeIndex.tz_localize](https://snowflakecomputing.atlassian.net/browse/SNOW-1017838?atlOrigin=eyJpIjoiMjZiMmY0MDQ0MTIwNGU0OWIyMzdjMWQ3ZDQwNTFkMDciLCJwIjoiaiJ9)
- [DatetimeIndex.tz_convert](https://snowflakecomputing.atlassian.net/browse/SNOW-1017838?atlOrigin=eyJpIjoiMjZiMmY0MDQ0MTIwNGU0OWIyMzdjMWQ3ZDQwNTFkMDciLCJwIjoiaiJ9)


Also:
- have to implement Timedelta tracking through `SnowflakeQueryCompiler::set_index_from_columns` (SNOW-1620436) to support the use case [here](https://github.com/snowflakedb/snowpark-python/blob/0a9bbc7a3e067cb97eac656453ac9e7ef733157d/tests/notebooks/modin/TimeSeriesTesting.ipynb#L265)

File: src/snowflake/snowpark/modin/plugin/_internal/binary_op_utils.py
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/_internal/binary_op_utils.py:246]:
```suggestion
def _is_timedelta_op(
```
nit

File: tests/notebooks/modin/MIMICHealthcareDemo.ipynb
Comment (sfc-gh-nkrishna) [reply tests/notebooks/modin/MIMICHealthcareDemo.ipynb:395]: @sfc-gh-mvashishtha Naresh's TimedeltaIndex PR supports that https://github.com/snowflakedb/snowpark-python/pull/2160

File: src/snowflake/snowpark/modin/plugin/_internal/binary_op_utils.py
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/_internal/binary_op_utils.py:246]: I actually prefer the original name `_op_is_between_two_timedeltas_or_timedelta_and_null` as I think it's clearer. `_is_timedelta_op` sounds ambiguous to me since it could be include an op between a Timedelta and a datetime which is not what we want here.

Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/_internal/binary_op_utils.py:246]: SG

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py
Comment (sfc-gh-mvashishtha) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:17375]: I think we'd have to update most of those places to handle timedelta-number comparisons correctly. We should add test coverage for each case, so I filed SNOW-1641716 to do that separately and I added a TODO pointing to that issue.

File: tests/integ/modin/test_timedelta_ops.py
Comment (sfc-gh-mvashishtha) [reply tests/integ/modin/test_timedelta_ops.py:34]: Done. I also deleted some redundant test cases in this file, since they seemed to just be doing the same date subtraction and timedelta division with different columns.

File: tests/notebooks/modin/MIMICHealthcareDemo.ipynb
Comment (sfc-gh-mvashishtha) [reply tests/notebooks/modin/MIMICHealthcareDemo.ipynb:395]: okay, I'll revisit that part after #2160 has been merged

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:17367]: sounds good.

**Pull request merged**
